packages/financial-templates-lib/src/price-feed/DefiPulseTotalPriceFeed.js

```diff
@@ -0,0 +1,144 @@
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { parseFixed } = require("@ethersproject/bignumber");
+const assert = require("assert");
+
+// An implementation of PriceFeedInterface that uses DefiPulse Data api to retrieve prices.
+class DefiPulseTotalPriceFeed extends PriceFeedInterface {
+  /**
+   * @notice Constructs the DefiPulseTVL_ALLPriceFeed.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} web3 Provider from truffle instance to connect to Ethereum network.
+   * @param {String} apiKey DeFiPulse Data API key. Note: these API keys are rate-limited.
+   * @param {Integer} lookback How far in the past the historical prices will be available using getHistoricalPrice.
+   * @param {Object} networker Used to send the API requests.
+   * @param {Function} getTime Returns the current time.
+   * @param {Integer} minTimeBetweenUpdates Min number of seconds between updates. If update() is called again before
+   *      this number of seconds has passed, it will be a no-op.
+   */
+  constructor(logger, web3, apiKey, lookback, networker, getTime, minTimeBetweenUpdates, decimals = 18) {
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+
+    this.apiKey = apiKey;
+    assert(apiKey, "requires apiKey for Defi Pulse");
+
+    this.lookback = lookback;
+    this.networker = networker;
+    this.getTime = getTime;
+    this.minTimeBetweenUpdates = minTimeBetweenUpdates;
+
+    this.toBN = this.web3.utils.toBN;
+
+    this.historicalPrices = [];
+
+    this.convertDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number
+      return this.toBN(parseFixed(number.toString(), decimals).toString());
+    };
+  }
+
+  getCurrentPrice() {
+    return this.currentPrice;
+  }
+
+  getHistoricalPrice(time) {
+    if (this.lastUpdateTime === undefined) {
+      return undefined;
+    }
+
+    let closestTime = { timestamp: 0, tvlUSD: 0 };
+
+    // go through all values and find time that that is the largest and still less than 'time'
+    for (let i = 0; i < this.historicalPrices.length; i++) {
+      let past = this.historicalPrices[i].timestamp;
+      let val = this.historicalPrices[i].tvlUSD;
+
+      if (past > closestTime.timestamp && past < time) {
+        closestTime.timestamp = past;
+        closestTime.tvlUSD = val;
+      }
+    }
+
+    const historicalPrice = this.web3.utils.toWei((closestTime.tvlUSD / 1000000000).toFixed(6), "ether");
```

Comment:
Does this respect the `decimals` value that is passed in the constructor?

Reply:
I removed decimals from the constructor.  The UMIP does specify 3 decimals of precision though so I made that change.  What I want is that when tvlUSD =  24781234567, the token price to be $24.781 or 24781000000000000000 wei of the collateral token.

Reply:
Am I understanding the price feed correctly? It should be returning a BN in wei correct?